executor/runcexecutor/executor.go

```diff
@@ -335,7 +335,11 @@ func (w *runcExecutor) Run(ctx context.Context, id string, root executor.Mount,
 
 	err = exitError(ctx, err)
 	if err != nil {
-		releaseContainer(context.TODO())
+		if rec != nil {
+			rec.CloseAsync(releaseContainer)
```

Comment:
I don't think we should change the release behavior depending on if recorder is enabled.

`CloseAsync` is an optimization to make sure the additional resource gathering and deleting cgroup by buildkit does not make container exec slower. But that case is not needed in case of error and I think the recorder should be just discarded synchronously instead of trying the sample it again async.

Reply:
Sounds good, I added a synchronous `Close` to the interface now and updated to call that in the error case.